### PR TITLE
Create Dockerfiles for Slim and Alpine builds

### PR DESCRIPTION
**Pull request by jeancochrane**
## Overview

Define two Dockerfile templates for Slim and Alpine builds, each of which compiles GDAL from source using Java and Python bindings. Bootstrap the Dockerfile templates off of https://github.com/geotrellis/geotrellis-gdal, with minor modifications to make builds faster.

Closes https://github.com/geotrellis/geotrellis-server/issues/60.

## Notes

* We had debated switching to an approach that would use the prebuilt [`libgdal-java`](https://packages.debian.org/stretch/libgdal-java) package to install these bindings. While this approach would benefit from faster build times and less maintenance on our end, I moved away from it for a few reasons:
    * `libgdal-java` is not available on `apk`
    * Nathan and Grigory both mentioned wanting to use the latest version of Geotrellis, which is only available on `experimental` Debian builds
    * After an hour or so of tweaking I couldn't get the package installation to pass the tests in https://github.com/geotrellis/geotrellis-server using a `sid-slim` build

## Testing instructions

* Confirm that Slim and Alpine builds compile successfully on Travis
* Run `GDAL_VERSION=2.3.2 OPENJPEG_VERSION=2.3.0 VARIANT=slim ./scripts/cibuild` to build a Slim image (note that this can take ~30 minutes, so plan ahead)
* Clone https://github.com/geotrellis/geotrellis-server and `cd` into it
* Run the following commands to test the images against real Geotrellis tests:

```
docker run \
  -v $(pwd)/build.sbt:/root/build.sbt \
  -v $(pwd)/project:/root/project \
  -v $(pwd)/core:/root/core \
  -v $(pwd)/sbt:/root/sbt \
  -w /root \
  quay.io/azavea/openjdk-gdal:2.3.2-slim \
  ./sbt "project core" test
```

* Clone https://github.com/geotrellis/geotrellis-gdal/ and `cd` into it
* Run the following commands to test the images against more real Geotrellis tests:

```
docker run \
  -v $(pwd)/build.sbt:/root/build.sbt \
  -v $(pwd)/project:/root/project \
  -v $(pwd)/gdal:/root/gdal \
  -v $(pwd)/src:/root/src \
  -v $(pwd)/sbt:/root/sbt \
  -w /root \
  quay.io/azavea/openjdk-gdal:2.3.2-slim \
  ./sbt "project gdal" test
```

### CODE REVIEW

File: Dockerfile-alpine.template
Comment (jeancochrane) [inline Dockerfile-alpine.template:6]: I'm not totally sure why we're compiling OpenJPEG from source, although I assume there's a version mismatch with GDAL. @moradology may have more context here.

File: Dockerfile-alpine.template
Comment (jeancochrane) [inline Dockerfile-alpine.template]: Ant is necessary to compile Java bindings, and is not available on `apk`.

File: Dockerfile-alpine.template
Comment (jeancochrane) [inline Dockerfile-alpine.template:46]: `make -j3` gives us up to three parallel jobs in Make. Xenial builds on Travis only have two cores available, but the docs mention [you can add extra](https://docs.travis-ci.com/user/speeding-up-the-build/#makefile-optimization) for heavy I/O (not sure how this actually works).

Note that the [GDAL CI pipeline uses `-j3` on Travis](https://github.com/OSGeo/gdal/blob/463e28a58d73e5f7bc4f554d999b2bb2e9bab799/gdal/ci/travis/ubuntu_1604/install.sh#L12).

File: Dockerfile-alpine.template
Comment (jeancochrane) [inline Dockerfile-alpine.template]: Not really sure why we need to `touch` this release file.

File: LICENSE
Comment (hectcastro) [inline LICENSE]: Reminder to update this to 2019 before merging.

File: scripts/cipublish
Comment (hectcastro) [inline scripts/cipublish]: I created a stub repository in Quay for this at https://quay.io/repository/azavea/openjdk-gdal. We'll still need to get the credentials for the `azavea+images` robot and feed them to Travis CI so that the login command above can succeed.

File: README.md
Comment (hectcastro) [inline README.md]: s/Geotrellis/GeoTrellis/

File: Dockerfile-slim.template
Comment (hectcastro) [inline Dockerfile-slim.template]: I think that we should try to make this more generic, like the Alpine image (e.g., `openjdk:8`).

As a side note, I tested this with a JRE based image vs. a JDK based one, and it fails to build (JRE image is usually a lot smaller, which is why we aim to use it when we can).

File: Dockerfile-alpine.template
Comment (hectcastro) [inline Dockerfile-alpine.template]: > Because image size matters, using `ADD` to fetch packages from remote URLs is strongly discouraged; you should use `curl` or `wget` instead. That way you can delete the files you no longer need after they’ve been extracted and you don’t have to add another layer in your image.

See: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy

I would consider doing this with something like:

```
wget -qO- http://download.osgeo.org/gdal/${GDAL_VERSION}/gdal-${GDAL_VERSION}.tar.gz \
| tar -xzC $ROOTDIR/src/ \
&& cd $ROOTDIR/src/
```

File: Dockerfile-alpine.template
Comment (hectcastro) [inline Dockerfile-alpine.template]: I think that this use of `apk` should be modified to remove build dependencies after we're done with them. Unfortunately, that might require sifting through these to determine which are true build dependencies vs. runtime ones.

https://github.com/azavea/docker-flask/blob/f9433b9a4d62336d9914b386ef9b178f93630c92/Dockerfile-alpine.template#L7-L12

(Note how the `pip install` step that needs to build dependencies is wrapped by the install and uninstall steps.)

File: Dockerfile-alpine.template
Comment (hectcastro) [inline Dockerfile-alpine.template]: It is unclear what this stage is for.

File: Dockerfile-alpine.template
Comment (hectcastro) [reply Dockerfile-alpine.template]: If @pomadchin doesn't know, then we should consider removing it.

File: Dockerfile-alpine.template
Comment (hectcastro) [inline Dockerfile-alpine.template]: Not sure this is something we want as the default `CMD`, especially for instances where we are using this to run a GeoTrellis service. I think it would be better to leave this stage out and bake it into `test`.

File: Dockerfile-alpine.template
Comment (hectcastro) [reply Dockerfile-alpine.template]: Could we use [`apache-ant`](https://pkgs.alpinelinux.org/package/edge/community/x86/apache-ant)?

File: Dockerfile-slim.template
Comment (hectcastro) [inline Dockerfile-slim.template]: Similar comment here as above regarding build dependencies:

https://github.com/azavea/docker-flask/blob/f9433b9a4d62336d9914b386ef9b178f93630c92/Dockerfile-slim.template#L7-L14

**Review by hectcastro — COMMENTED** [review]
Left most of my comments on the Alpine `Dockerfile`, but a bunch apply to the Debian one too.

File: Dockerfile-alpine.template
Comment (jeancochrane) [reply Dockerfile-alpine.template]: That's awesome context 👍

File: Dockerfile-slim.template
Comment (jeancochrane) [reply Dockerfile-slim.template]: Do you think it's worth trying to get the JRE image to build? Seems worth a try, but it'll likely take some time to figure out.

Comment (hectcastro) [reply Dockerfile-slim.template]: Probably not. It may be doable, but there seem to be enough other fish to fry here.

File: scripts/cipublish
Comment (jeancochrane) [reply scripts/cipublish]: I temporarily set Travis to deploy from this branch in https://github.com/azavea/docker-openjdk-gdal/pull/1/commits/85a30ec017038408e2d4ec89a57d272a95b93006 to make sure that this would work. Images are up here: https://quay.io/repository/azavea/openjdk-gdal The commit has been reverted, so we should only deploy from `master` now.

**Comment by jeancochrane** [comment]
Alright, this should be ready for another pass.

File: scripts/cibuild
Comment (hectcastro) [inline scripts/cibuild]: I think that we need to make the tag a bit more descriptive, which could mean that we need to parametrize OpenJDK version, but I could also see just leaving it as is, but also hardcoding the version into the tag name like:

```
quay.io/azavea/openjdk-gdal:8-jdk-gdal-2.3
```

Not pretty, but I think we need a way to surface both the underlying JDK and GDAL versions.

File: Dockerfile-alpine.template
Comment (hectcastro) [inline Dockerfile-alpine.template]: I think that you can use the `--update` flag for `apk add` vs. having to do a standalone `update` here.

**Review by hectcastro — COMMENTED** [review]
Changes here all look good, but now my `geotrellis-gdal` tests are failing. One of the output messages is:

```
[info] - Java GDAL bindings not installed, skipping !!! IGNORED !!!
```

Failures are happening for both the `slim` and `alpine` variants.

File: scripts/test
Comment (hectcastro) [inline scripts/test]: I think that this would be a valuable spot to encode something like:

```
gdalinfo --version
gdalinfo --formats
ogrinfo --formats
```

File: scripts/cibuild
Comment (jeancochrane) [reply scripts/cibuild]: How about:

```
quay.io/azavea/openjdk-gdal:2.3.2-jdk8-slim
```

So that we can leverage the implied GDAL versioning in `gdal:2.3.2`? That seems to be similar to the pattern we follow with our Django images: https://quay.io/repository/azavea/django?tag=latest&tab=tags

Comment (hectcastro) [reply scripts/cibuild]: Heh, I went back-and-forth on it and settled on putting the JDK version first because it is first in the container image name. I also used `8-jdk` to align with the official container image naming structure.

This might be a thing where you propose both options to 3 people, see what makes sense to the majority, and just go with that.

**Review by hectcastro — APPROVED** [review]
I think that this is in a good spot. Naming stuff doesn't need to be sorted out before approval, and I was able to get the `geotrellis-gdal` tests passing after tweaking the GDAL version reference, like you mentioned. 👍

**Pull request merged**
